File: backend/app/schemas/employee.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field


# ============================================================
# CREATE EMPLOYEE
# ============================================================

class EmployeeCreate(BaseModel):

    employee_code: str = Field(
        ...,
        min_length=3,
        max_length=20
    )

    name: str = Field(
        ...,
        min_length=2,
        max_length=100
    )

    phone: str = Field(
        ...,
        min_length=7,
        max_length=20
    )

    email: EmailStr

    branch_id: int = Field(
        ...,
        gt=0
    )

    vehicle_type: str = Field(
        ...,
        min_length=2,
        max_length=50
    )

    status: str = Field(
        default="ACTIVE",
        max_length=30
    )

    current_latitude: float | None = Field(
        default=None,
        ge=-90,
        le=90
    )

    current_longitude: float | None = Field(
        default=None,
        ge=-180,
        le=180
    )


# ============================================================
# UPDATE EMPLOYEE
# ============================================================

class EmployeeUpdate(BaseModel):

    name: str | None = Field(
        default=None,
        min_length=2,
        max_length=100
    )

    phone: str | None = Field(
        default=None,
        min_length=7,
        max_length=20
    )

    email: EmailStr | None = None

    branch_id: int | None = Field(
        default=None,
        gt=0
    )

    vehicle_type: str | None = Field(
        default=None,
        min_length=2,
        max_length=50
    )

    status: str | None = Field(
        default=None,
        max_length=30
    )

    current_latitude: float | None = Field(
        default=None,
        ge=-90,
        le=90
    )

    current_longitude: float | None = Field(
        default=None,
        ge=-180,
        le=180
    )


# ============================================================
# EMPLOYEE RESPONSE
# ============================================================

class EmployeeResponse(BaseModel):

    id: int

    employee_code: str

    name: str

    phone: str

    email: str

    branch_id: int

    vehicle_type: str

    status: str

    current_latitude: float | None

    current_longitude: float | None

    total_deliveries: int

    completed_deliveries: int

    delayed_deliveries: int

    average_delivery_time: float

    performance_score: float

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# EMPLOYEE LIST RESPONSE
# ============================================================

class EmployeeListResponse(BaseModel):

    total: int

    page: int

    limit: int

    employees: list[EmployeeResponse]

File: frontend/src/types/employee.ts
export interface Employee {
  id: number;
  employee_code: string;
  name: string;
  phone: string;
  email: string;
  branch_id: number;
  vehicle_type: string;
  status: "ACTIVE" | "INACTIVE";
  current_latitude: number | null;
  current_longitude: number | null;
  total_deliveries: number;
  completed_deliveries: number;
  delayed_deliveries: number;
  average_delivery_time: number;
  performance_score: number;
  created_at: string;
  updated_at: string;
}

export interface EmployeeCreate {
  employee_code: string;
  name: string;
  phone: string;
  email: string;
  branch_id: number;
  vehicle_type: string;
  status: "ACTIVE" | "INACTIVE";
  current_latitude: number | null;
  current_longitude: number | null;
}

export interface EmployeeUpdate {
  name?: string;
  phone?: string;
  email?: string;
  branch_id?: number;
  vehicle_type?: string;
  status?: "ACTIVE" | "INACTIVE";
  current_latitude?: number | null;
  current_longitude?: number | null;
}

export interface EmployeeListResponse {
  total: number;
  page: number;
  limit: number;
  employees: Employee[];
}

export interface Branch {
  id: number;
  branch_code?: string;
  branch_name: string;
  address?: string;
  city?: string;
  latitude?: number | null;
  longitude?: number | null;
  phone?: string;
  status?: string;
}